src/scratch_comment_icon.js

```diff
@@ -0,0 +1,221 @@
+import * as Blockly from 'blockly/core';
+import {ScratchCommentBubble} from './scratch_comment_bubble.js';
+
+class ScratchCommentIcon extends Blockly.icons.Icon {
+  constructor(sourceBlock) {
+    super(sourceBlock);
+    this.sourceBlock = sourceBlock;
+    this.commentBubble = new ScratchCommentBubble(this.sourceBlock, this.getAnchorPoint());
+    Blockly.Events.setGroup(true);
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_create'))(
+        this.commentBubble
+      ),
+    );
+    this.onTextChangedListener = this.onTextChanged.bind(this);
+    this.onSizeChangedListener = this.onSizeChanged.bind(this);
+    this.onCollapseListener = this.onCollapsed.bind(this);
+    this.commentBubble.addTextChangeListener(this.onTextChangedListener);
+    this.commentBubble.addSizeChangeListener(this.onSizeChangedListener);
+    this.commentBubble.addOnCollapseListener(this.onCollapseListener);
+    this.repositionAfterRender = true;
+
+    // We need to force a render of our parent block and wait for it to complete, because at this
+    // point, we don't know our own position (needed to calculate the anchor point and position
+    // the comment bubble properly relative to the block) and our parent block doesn't know if it's
+    // an insertion marker or not (needed to suppress showing the comment bubble attached to an
+    // insertion marker block).
+    this.sourceBlock.queueRender();
+    Blockly.renderManagement.finishQueuedRenders().then(() => {
+      if (!this.sourceBlock || !this.commentBubble) return;
+
+      if (this.sourceBlock.isInsertionMarker()) {
+        this.commentBubble.dispose();
+        return;
+      }
+
+      // loadFromState may have already positioned the bubble appropriately when e.g. populating a
+      // comment from the undo stack. Only position the bubble at the default location if
+      // loadFromState hasn't prohibited repositioning post-render.
+      if (this.repositionAfterRender) {
+        const anchor = this.getAnchorPoint();
+        this.commentBubble.moveTo(anchor.x + 40, anchor.y - 16);
+      }
+      Blockly.Events.setGroup(false);
+    });
+  }
+
+  getType() {
+    return Blockly.icons.IconType.COMMENT;
+  }
+
+  initView(pointerDownListener) {
+    // Scratch comments have no indicator icon on the block.
+    return;
+  }
+
+  getSize() {
+    // Awful hack to cancel out the default padding added to icons.
+    return new Blockly.utils.Size(-8, 0);
+  }
+
+  getAnchorPoint() {
+    const sourceBlockOrigin = this.sourceBlock.getRelativeToSurfaceXY();
+    let left;
+    let right;
+    if (this.sourceBlock.RTL) {
+      left = sourceBlockOrigin.x - this.sourceBlock.width;
+      right = sourceBlockOrigin.x;
+    } else {
+      left = sourceBlockOrigin.x;
+      right = sourceBlockOrigin.x + this.sourceBlock.width;
+    }
```

Comment:
This looks like it's in workspace coordinates? in which case you can use `getBoundingRectangle`

Reply:
`getBoundingRectangle` doesn't work because it includes subsequent blocks, and we need the actual rect of this block alone to determine the vertical midpoint for the anchor line to emanate from.

Reply:
Updated this to use the newly introduced `getBoundingRectangleWithoutChildren()`.